### ISSUE TITLE
Conditional ownership for interfacing wrappers

### ISSUE THREAD

**Issue by nbeams** [ginkgo-project/ginkgo#444]
**Background**

When creating wrappers for interfacing with exterior libraries' data structures via pointers, we seemingly have a situation where we want "conditionally owning" pointers for the exterior library data.  That is, sometimes we want the Ginkgo wrapper object to own the data, and sometimes we just want to point to the data but leave its management entirely up to the exterior library.  This issue is for discussion of how we would want to go about doing this.

An example of an MFEM Vector wrapper using plain pointers is in the mfem-wrappers branch, in examples/mfem-wrappers/mfem_wrapper.hpp.  The original intention of the wrapper was how it is used in test_mfem.cpp, e.g.:

```
// MFEM code:
Vector B, X;
a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);

// At this point, X and B are ready for MFEM to use, so we just want
// Ginkgo to point to the data, and create the wrapper this way:
 
auto gko_rhs = MFEMVectorWrapper::create(executor, B.Size(), &B);
auto gko_x = MFEMVectorWrapper::create(executor, X.Size(), &X);
```
We don't really want the wrappers to have an owning pointer to B or X, because, as I found,
you can have issues with garbage collection with the top-level code that uses both `X` and `gko_x`. (There are also cases with MFEM where the pointer `*b` and Vector `B` end up pointing to the same memory, which could also easily cause deallocation problems.)

However, another necessary piece to make these wrappers work with Ginkgo solvers was the 
`create_with_same_config()` function, which I've currently implemented as
```
    virtual std::unique_ptr<gko::matrix::Dense<double>>
    create_with_same_config() const override
    {
        mfem::Vector *mfem_vec = new mfem::Vector(this->get_size()[0]);
        return MFEMVectorWrapper::create(this->get_executor(),
                                         this->get_size()[0], mfem_vec);
    }
```
In this case, we've created a new MFEM Vector inside Ginkgo, which will only be used temporarily by the Ginkgo solvers.  The top-level MFEM user probably would never call this code (the current example test_mfem.cpp does not).  This memory could/should be owned by the wrapper class to avoid memory leaks.

**The question**

What is the best solution for conditionally-owning pointers?  Ownership flags and a specialized destructor for the wrapper class?  Is there something more "`C++`y" or less prone to user error?  How would a "top-level" user (e.g. someone writing MFEM code) interact with the ownership issue?

**Comment by upsj** [comment]
Slightly related: We already do something similar in `gko::Array`:
The [`view` factory function](https://github.com/ginkgo-project/ginkgo/blob/d53f39bee3c4d973be61f1dd42cdeda5fb0e1662/include/ginkgo/core/base/array.hpp#L254) allows for creating non-owning arrays. This works by using a `std::unique_ptr` with a polymorphic `std::function<void(T*)>` deleter.

**Comment by nbeams** [comment]
My latest commit (68a4889d28baa18f44f3fc21a77c9f8cfb4164ab) on the mfem-wrappers branch is a first attempt at this. In this version, an ownership flag is passed in upon creation, but not stored or used anywhere else. The default value of the flag is `false`, so if the user doesn't provide it, it will assume MFEM has already created and should maintain the memory.  The `unique_ptr` to the MFEM Vector object is created with one of two deleters based on the ownership flag:

1. Thanks to @upsj's comment, I used the Ginkgo `null_deleter` for the non-owning case, and created my own version for the owning case.  

2. Originally, I tried the default destructor for the owning case, but I was still getting errors upon deallocation.  I think the problem is that `~Vector()` in MFEM only destroys the Vector's owned pointers, and doesn't free the underlying `Memory` object.  The `Destroy` routine seems to do both (though I haven't yet done checks to see if I have memory leaks).
So that means the MFEM deleter is 
```
template <typename T>
class mfem_destroy {
public:
    using pointer = T *;

    /**
     * Destroys an MFEM object.  Requires object to have a Destroy() method.
     *
     * @param ptr  pointer to the object being deleted
     */
    void operator()(pointer ptr) const noexcept { ptr->Destroy(); }
};
```
Inside the constructor, I have 
```
    if (ownership) {
            using deleter = mfem_destroy<mfem::Vector>;
            mfem_vec_ = std::unique_ptr<mfem::Vector,
                                        std::function<void(mfem::Vector *)>>(
                mfem_vec, deleter{});
        } else {
            using deleter = gko::null_deleter<mfem::Vector>;
            mfem_vec_ = std::unique_ptr<mfem::Vector,
                                        std::function<void(mfem::Vector *)>>(
                mfem_vec, deleter{});
        }
```
to set the private pointer variable in the wrapper.

Then the `create_with_same_config` calls the create method with `true` for the ownership [here](https://github.com/ginkgo-project/ginkgo/blob/68a4889d28baa18f44f3fc21a77c9f8cfb4164ab/examples/mfem-wrappers/mfem_wrapper.hpp#L93).

This probably isn't the final version of what we want, but at least it does appear to work (again, pending memory leak issues I haven't found yet).

**Comment by nbeams** [comment]
Well, I finally seemingly found something that fixed the weird memory deallocation error when using cuda, but I'm confused.  

As mentioned above, using the null deleter for the non-owning case fixed the memory deallocation error in the CPU version, preventing two objects from trying to delete the memory.  However, even using this null deleter version, adding `gko_x.release()` at the end of the `test_mfem.cpp` program (just before the other MFEM memory deletions) finally removed the cudaFree error.  Also, when I ran the version of the test with the `release` call through `cuda-memcheck --leak-check full`, it exited successfully and reported 0 memory leaks.

However, I don't understand why this behavior should be different for cuda vs CPU, and require the explicit disowning of the smart pointer in only one case. 

One thing I noticed:  in the test, I wrap both `x` and `b`.  `x` is an MFEM GridFunction and `b` is an MFEM LinearForm, both of which are derived from Vector (and in this test problem, end up pointing to the same memory as Vectors `X` and `B`).  Only `x` needs the `release()` call.  By the end of the program, `b` is valid only on the host (as reported by MFEM's Memory class flags), while `x` is valid on host and device.  But still, if the `unique_ptr`'s deleter doesn't actually delete anything, I don't see why it would matter for the cuda case.

**Comment by thoasm** [comment]
@nbeams Is the pushed version in [`mfem-wrappers`](https://github.com/ginkgo-project/ginkgo/tree/mfem-wrappers) up-to-date (like the file [`examples/mfem-wrappers/mfem_wrappers.hpp`](https://github.com/ginkgo-project/ginkgo/blob/mfem-wrappers/examples/mfem-wrappers/mfem_wrapper.hpp))? It says last updated 13 days ago.
The `unique_ptr` should take care of the problem like you said, but I would need to look at the code to see the origin of the problem.

**Comment by nbeams** [comment]
No, I need to clean up some (okay, a lot) of test/commented out code and update the branch. I'll work on that now.

What's really weird is that yesterday, I commented out the `release()` call to try to reproduce the errors I was getting before, and now it worked fine without it...?  So I'm even more confused.

**Comment by nbeams** [comment]
Okay, 63080a5806bb34e434ffb286579f832ce98fe917 is my updated code.  The problems without `release()` are back again today.  Maybe I didn't do a `make clean` when I should have or something.  Also, I still can't get it to run on other GPUs than the V100. 

In this version I have introduced for testing purposes an explicit parameter for `on_device` in the `create` function for the VectorWrapper.  I'm pretty sure I am going to remove this again.

I took out the extra comments and such, but a useful thing to see what MFEM is doing, for a Vector or derived class, is to call `GetMemory().PrintFlags()` to see the Memory object's current status.
